numpy/core/src/umath/simd.inc.src

```diff
@@ -55,6 +55,40 @@ abs_ptrdiff(char *a, char *b)
     return (a > b) ? (a - b) : (b - a);
 }
 
+/*
+ * nomemoverlap - Do two strided arrays overlap?
+ * Need to accomodate negative strides.
```

Comment:
Did you accommodate negative strides? The question is a bit tricky, because it used to be the case that they can overlap, however that should not happen anymore (but this is ensured in a completely different part of the code).

Unless my coffee has not kicked in, reductions also have an exact overlap. So basically, it should be sufficient to ensure exact overlapping. But, we should document that somewhere that some inner loops rely on that, and I am not sure where.

Reply:
Oh, two strided arrays definitely can overlap, but I think only identically so.

Reply:
My bad, the function does already accommodate negative strides, I will re-write the comment :)

Reply:
The loops can be called with overlapping arrays if the data dependence is simple, assuming elementwise operations. This I think currently means in practice that overlap can occur in the reduction/accumulation usage, and when the output pointer never catches the inputs when iterated in order.

Vectorized loops have more strict requirement, so I think they probably should do their own checks.

Reply:
The function `nomemoverlap `returns a false if the two arrays have overlap. It accommodates for negative strides too. The vectorized loops rely on this function to ensure there is absolutely no overlap between input and output array, so in that sense it is being overly cautious.